datastream/tools/split_dataframe.py

```diff
@@ -0,0 +1,155 @@
+from __future__ import annotations
+from typing import Tuple, Union, Dict, Optional
+from pathlib import Path
+import numpy as np
+import pandas as pd
+
+from datastream import json
+
+
+def split_dataframe(
+    dataframe: pd.DataFrame,
+    key_column: str,
+    proportions: Dict[str, float],
+    filepath: Optional[Path] = None,
+    stratify_column: Optional[str] = None,
+):
+    '''
+    Split and save result. Add new examples and continue from the old split.
+
+    As new examples come in it can handle:
+    - Changing test size
+    - Adapt after removing examples from dataset
+    - Adapt to new stratification
+    '''
+    if abs(sum(proportions.values()) - 1.0) >= 1e-6:
+        raise AssertionError('Expected sum of proportions to be 1')
+    
+    if filepath is not None and filepath.exists():
+        split = json.read(filepath)
+
+        if set(proportions.keys()) != set(split.keys()):
+            raise AssertionError(
+                'Expected split names in split file to be the same as the keys'
+                'of proportions'
+            )
+    else:
+        split = {
+            split_name: list()
+            for split_name in proportions.keys()
+        }
+
+    split_proportions = tuple(proportions.items())
+
+    for split_name, proportion in split_proportions[:-1]:
+        if stratify_column is None:
+            split = split_proportion(
+                dataframe[key_column],
+                proportion,
+                split_name,
+                split,
+            )
+        else:
+            for strata in stratas(dataframe, stratify_column):
+                split = split_proportion(
+                    dataframe[key_column],
+                    proportion,
+                    split_name,
+                    split,
+                )
+
+    last_split_name, _ = split_proportions[-1]
+    split[last_split_name] += unassigned(dataframe[key_column], split)
+
+    if filepath is not None:
+        json.write(split, filepath)
+
+    return {
+        split_name: (
+            dataframe[lambda df: df[key_column].isin(split[split_name])]
+        )
+        for split_name in proportions.keys()
+    }
+
+
+def stratas(dataframe, stratify_column):
+    return [
+        strata_value for strata_value, _ in (
+            dataframe[stratify_column].value_counts().iteritems()
```

Comment:
Alternatively
```python 
return [strata for strata in dataframe[stratify_column].unique()]
```

Reply:
Could also have this function return the stratas directly instead of just the strata values?
```
return [
    dataframe[dataframe[stratify_column] == strata_value]
    for strata_value in dataframe[stratify_column].unique()
]
```

Reply:
Just copied from the old code. I found a bug thanks to this. Thanks! :)